server/src/controllers/authentication_controller.ts

```diff
@@ -1,44 +1,95 @@
 import "reflect-metadata";
-import * as restify from "restify";
+import { Request, Response } from "restify";
 import { Controller, interfaces, Post } from "inversify-restify-utils";
 import { inject, injectable, named } from "inversify";
 import { AuthenticationService } from "../services/authentication_service";
 import * as Joi from "joi";
+import { GENDERS } from "../entities/gender";
+import { RequestUser } from "../entities/request/RequestUser";
+import { RequestAddress } from "../entities/request/RequestAddress";
 
 @Controller("/")
 @injectable()
 export class AuthenticationController implements interfaces.Controller {
     constructor(
         @inject("Service")
         @named("AuthenticationService")
-        private readonly _authenticationService: AuthenticationService,
+        private readonly authenticationService: AuthenticationService,
     ) {}
 
-    @Post("/signup")
-    private async signUp(req: restify.Request, res: restify.Response): Promise<any> {
-        const schema: Joi.ObjectSchema = Joi.object({
-            email: Joi.string().email(),
-            // Password requires upper case and lower case letter as well as number and special character.
-            password: Joi.string()
-                .pattern(new RegExp("^(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z])(?=.*[!@#$%^&*()?+=_]).{8,30}$"))
-                .required(),
-            first_name: Joi.string().pattern(new RegExp("^[A-Za-z]+")).min(2).max(25).required(),
-            last_name: Joi.string().pattern(new RegExp("^[A-Za-z]+")).min(2).max(25).required(),
-            date_of_birth: Joi.date().required(),
-            created_on: Joi.date().timestamp(),
-        });
-
-        const { error } = await schema.validateAsync(req.body);
-
-        if (!req.is("application/json") || error) {
-            // indicate error
-            return res.send(400, error || "Data is not in json format.");
+    @Post("/sign_up")
+    private async signUp(req: Request, res: Response): Promise<void> {
+        const { value, error } = signUpSchema.validate(req.body);
+
+        if (error) {
+            res.json(400, error);
+            return;
         }
 
-        const e = await this._authenticationService.createUser(req.body);
-        if (!e) {
-            return res.send(200, "New user has been created.");
+        const userData: RequestUser = {
```

Comment:
Just a suggestion to maybe move this code in a service class to shorten the method length.

Reply:
So we are following a layered architecture
controller -> responsible for handling user input which has 3 components
- routing
- data validation and transformation
- handling the response

Moving data validation into the service layer breaks the separation of layers because you pollute the service layer (business logic) with semantics of the request. The idea is once we hit the service layer the data is validated and in the correct format.